File: python/1774.py
from collections import defaultdict

if __name__ == "__main__":
    r, c = map(int, input().split(" "))

    dag: dict[int, list[tuple[int, int]]] = defaultdict(list)

    for ci in range(c):
        a, b, p = map(int, input().split(" "))
        dag[a].append((b, p))
        dag[b].append((a, p))

    visited = set()
    visited.add(1)

    nexts: set[tuple[int, int]] = set()
    for i in dag[1]:
        nexts.add(i)

    current_tree = []

    while len(visited) != len(dag.keys()):
        # next node with min value
        next = min(nexts, key=lambda a: a[1])
        nexts.remove(next)

        # if this node was already visited, we discard this iteration
        if next[0] in visited:
            continue

        # we add all conected node to the list
        for i in dag[next[0]]:
            nexts.add(i)

        # since this node is the one with lower value,
        # we add it to the current_tree
        current_tree.append(next[1])

        # we mark this node as visited and
        # continue the iteration until visit all nodes
        visited.add(next[0])

    print(sum(current_tree))

File: go/1774.go
package main

import (
	"fmt"
)

type Node struct {
	weight int
	value  int
}

func main() {
	var r, c int

	fmt.Scanf("%d %d", &r, &c)

	dag := make(map[int][]Node)

	for i := 0; i < c; i++ {
		var a, b, p int
		fmt.Scanf("%d %d %d", &a, &b, &p)
		dag[a] = append(dag[a], Node{weight: p, value: b})
		dag[b] = append(dag[b], Node{weight: p, value: a})
	}

	visited_count := 1
	visited := make(map[int]bool)
	visited[1] = true

	nexts := []Node{}
	nexts = append(nexts, dag[1]...)

	total := 0
	for range dag {
		total += 1
	}

	shortest := 0

	for visited_count < total {
		var next Node
		var idx int
		for i, ni := range nexts {
			if ni.weight < next.weight || next.value == 0 {
				next = ni
				idx = i
			}
		}

		nexts = append(nexts[:idx], nexts[idx+1:]...)

		if visited[next.value] {
			continue
		}

		nexts = append(nexts, dag[next.value]...)

		shortest += next.weight

		visited[next.value] = true
		visited_count += 1
	}

	fmt.Printf("%d\n", shortest)

}
